### PR TITLE
Added api for putting state via cli/rest

### PR DESCRIPTION
**Pull request by srkukarni**
### Motivation
Currently state can be altered inside the function via the context.setState. There is no other mechanism to alter this state outside the function. This pr adds a rest interface and the corresponding cli handler for putState
### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by srkukarni** [comment]
run java8 tests

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java [hunk 1]
```diff
@@ -830,6 +831,20 @@ void runCmd() throws Exception {
...
 
+    @Parameters(commandDescription = "Put the state associated with a Pulsar Function")
+    class StatePutter extends FunctionCommand {
+
+        @Parameter(names = { "-s", "--state" }, description = "The FunctionState that needs to be put", required = true)
+        private String state = null;
+
+        @Override
+        void runCmd() throws Exception {
+            FunctionState stateRepr = new Gson().fromJson(state, FunctionState.class);
```
Comment (jerrypeng) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java]: Lets ObjectMapperFactory.getThreadLocal() instead Gson() its more efficient

File: pulsar-common/src/main/java/org/apache/pulsar/common/functions/FunctionState.java [hunk 2]
```diff
@@ -34,6 +34,7 @@
 public class FunctionState {
     private String key;
     private String stringValue;
+    private byte[] byteValue;
```
Comment (jerrypeng) [inline pulsar-common/src/main/java/org/apache/pulsar/common/functions/FunctionState.java]: Is this relevant to this PR?

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java [hunk 3]
```diff
@@ -1470,9 +1470,13 @@ public FunctionState getFunctionState(final String tenant,
                     throw new RestException(Status.NOT_FOUND, "key '" + key + "' doesn't exist.");
                 } else {
                     if (kv.isNumber()) {
-                        value = new FunctionState(key, null, kv.numberValue(), kv.version());
+                        value = new FunctionState(key, null, null, kv.numberValue(), kv.version());
                     } else {
-                        value = new FunctionState(key, new String(ByteBufUtil.getBytes(kv.value()), UTF_8), null, kv.version());
+                        try {
+                            value = new FunctionState(key, new String(ByteBufUtil.getBytes(kv.value(), kv.value().readerIndex(), kv.value().readableBytes()), UTF_8), null, null, kv.version());
```
Comment (jerrypeng) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java]: is this relevant to this PR?

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java [hunk 4]
```diff
@@ -1490,6 +1494,87 @@ public FunctionState getFunctionState(final String tenant,
...
+        } catch (PulsarAdminException e) {
+            log.error("{}/{}/{} Failed to authorize [{}]", tenant, namespace, functionName, e);
+            throw new RestException(Status.INTERNAL_SERVER_ERROR, e.getMessage());
+        }
+
+        if (null == worker().getStateStoreAdminClient()) {
+            throwStateStoreUnvailableResponse();
+        }
+
+        FunctionState state = new Gson().fromJson(stateJson, FunctionState.class);
```
Comment (jerrypeng) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java]: Please use ObjectMapperFactory.getThreadLocal() instead of creating anew Gson object

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java [hunk 5]
```diff
@@ -1490,6 +1494,87 @@ public FunctionState getFunctionState(final String tenant,
...
+                log.error("{}/{}/{} Client [{}] is not admin and authorized to put state for {}", tenant, namespace,
+                        functionName, clientRole, ComponentTypeUtils.toString(componentType));
+                throw new RestException(Status.UNAUTHORIZED, "client is not authorize to perform operation");
+            }
+        } catch (PulsarAdminException e) {
+            log.error("{}/{}/{} Failed to authorize [{}]", tenant, namespace, functionName, e);
+            throw new RestException(Status.INTERNAL_SERVER_ERROR, e.getMessage());
+        }
+
+        if (null == worker().getStateStoreAdminClient()) {
```
Comment (jerrypeng) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java]: This check makes more sense to be done after checking isWorkerServiceAvailable() and before isAuthorizedRole()

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java [hunk 6]
```diff
@@ -1490,6 +1494,87 @@ public FunctionState getFunctionState(final String tenant,
...
+
+        ByteBuf value;
+        if (!isEmpty(state.getStringValue())) {
+            value = Unpooled.wrappedBuffer(state.getStringValue().getBytes());
+        } else {
+            value = Unpooled.wrappedBuffer(state.getByteValue());
+        }
+        try (Table<ByteBuf, ByteBuf> table = result(storageClient.get().openTable(tableName))) {
+            result(table.put(Unpooled.wrappedBuffer(key.getBytes(UTF_8)), value));
+        } catch (RestException e) {
```
Comment (jerrypeng) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java]: why are we catching an unchecked exception and then throwing it again without doing anything else?  Seems unnecessary

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java (hunk 1 above)
Comment (srkukarni) [reply pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java]: changed

File: pulsar-common/src/main/java/org/apache/pulsar/common/functions/FunctionState.java (hunk 2 above)
Comment (srkukarni) [reply pulsar-common/src/main/java/org/apache/pulsar/common/functions/FunctionState.java]: For both get and put state, we need to know if we are dealing with byte values or string. The current version only assumes string. Hence the change

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java (hunk 3 above)
Comment (srkukarni) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java]: same as above

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java (hunk 4 above)
Comment (srkukarni) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java]: done

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java (hunk 5 above)
Comment (srkukarni) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java]: changed

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java (hunk 6 above)
Comment (srkukarni) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/ComponentImpl.java]: changed

**Comment by srkukarni** [comment]
@jerrypeng added tests as well

**Comment by srkukarni** [comment]
One thing that I’m using here is functionstate class for getting both the state stored as well as for changing the state. Another approach would be to have an alternate structure for changing state. What do you feel?

**Comment by srkukarni** [comment]
run java8 tests
run integration tests

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/FunctionsBase.java [hunk 7]
```diff
@@ -276,6 +276,26 @@ public FunctionState getFunctionState(final @PathParam("tenant") String tenant,
...
+            @ApiResponse(code = 404, message = "The function does not exist"),
+            @ApiResponse(code = 500, message = "Internal server error")
+    })
+    @Path("/{tenant}/{namespace}/{functionName}/state/{key}")
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    public void putFunctionState(final @PathParam("tenant") String tenant,
+                                 final @PathParam("namespace") String namespace,
+                                 final @PathParam("functionName") String functionName,
+                                 final @PathParam("key") String key,
+                                 final @FormDataParam("state") String stateJson) {
```
Comment (jerrypeng) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/FunctionsBase.java]: just this as FunctionState, you don't need to get it as a string.  Jersey will automatically deserialize that for you

Comment (jerrypeng) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/FunctionsBase.java]: I don't think we should use "FormData", that is usually used for submitting via forms on a website.  You just have "FunctionState state" as input here

Comment (jerrypeng) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/FunctionsBase.java]: I know we use "FormData" everything where else but I don't think that is correct

**Comment by jerrypeng** [comment]
> One thing that I’m using here is functionstate class for getting both the state stored as well as for changing the state. Another approach would be to have an alternate structure for changing state. What do you feel?

@srkukarni I think that is fine for now

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/FunctionsBase.java (hunk 7 above)
Comment (srkukarni) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/FunctionsBase.java]: Done

Comment (srkukarni) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/FunctionsBase.java]: What will happen if we include more items in the future? How will it know which keys map to which data?

Comment (jerrypeng) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/FunctionsBase.java]: In that case, the input should be a json object that contains all the fields.  That is usually what people are doing nowadays.  For example:

https://www.vaultproject.io/api/secret/databases/cassandra.html#sample-payload

Comment (srkukarni) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/FunctionsBase.java]: Take the example of 'big objects'. In future versions we might introduce a way to stream the big object via an inputstream. Thus a single object might not work for this case?

**Comment by srkukarni** [comment]
@jerrypeng ^^^

**Comment by srkukarni** [comment]
run integration tests

**Comment by srkukarni** [comment]
run integration tests
